### PR TITLE
Document the current process for merging from FreeBSD

### PR DESCRIPTION
**Pull request by brooksdavis**

### CODE REVIEW

File: CHERI-UPDATING.md [hunk 1]
```diff
@@ -79,6 +79,152 @@ Each merge from upstream FreeBSD to [dev] is accompanied by a tag of the form
...
+#### Finishing up
+Once the merge is complete and basic local testing is done, it's time to
+update the `freebsd-main` branch and push the tag. (It's good to delay
+to this point in hopes of keeping `freebsd-main` relatively in sync with
+main.)
+```
+$ git push origin freebsd-main-<YYYYMMDD>:freebsd-main
+$ git push origin freebsd-main-<YYYYMMDD>
+```
+Now, push the main tree to create a pull request for kick off a full CI run.
```
Comment (bsdjhb) [inline CHERI-UPDATING.md:198]:
```suggestion
Now, push the main tree to create a pull request to kick off a full CI run.
```

**Review by bsdjhb — APPROVED** [review]
This looks fairly clear to me, thanks a bunch!

File: CHERI-UPDATING.md [hunk 2]
```diff
@@ -79,6 +79,152 @@ Each merge from upstream FreeBSD to [dev] is accompanied by a tag of the form
...
+```
+$ git remote add upstream https://git.FreeBSD.org/src.git
+```
+You will also want to enable `git revere` to simplify handling the same
+conflicting merges over again. To do so globally, run:
+```
+$ git config --global rerere.enabled true
+```
+Disable rename detection. This sometimes leads to awkward merges but
+rename detection takes **forever** on the CheriBSD repo.
```
Comment (nwf) [inline CHERI-UPDATING.md:107]: Want to give the command?

File: CHERI-UPDATING.md [hunk 3]
```diff
@@ -79,6 +79,152 @@ Each merge from upstream FreeBSD to [dev] is accompanied by a tag of the form
...
+a few commits up or down to make sure the commit isn't at an obviously
+bad point. Then we create a tag for the commit:
+```
+$ git tag -m "FreeBSD main DD Month YYYY" freebsd-main-<YYYYMMDD> <hash>
+```
+
+#### Performing the merge
+First, create a merge branch from an update to date `dev` checkout:
+```
+$ git checkout -b merge-freebsd-<YYYYMMDD>
```
Comment (nwf) [inline CHERI-UPDATING.md:139]: I think you can use
```suggestion
$ git checkout -b merge-freebsd-<YYYYMMDD> dev
```
to help ensure that you start from the right place, regardless of the current checkout?

File: CHERI-UPDATING.md [hunk 4]
```diff
@@ -79,6 +79,152 @@ Each merge from upstream FreeBSD to [dev] is accompanied by a tag of the form
...
+use `git reset` to rewind to the broken commit, fix it up, and restart
+the merge. The process usually looks like
+```
+$ mergify abort
+$ git reset --hard <bad merge>
+...
+$ git commit --amend
+$ mergify start freebsd-main-<YYYYMMDD>
+```
+Sometime the commit it self isn't the problem, but rather that it
```
Comment (nwf) [inline CHERI-UPDATING.md:178]:
```suggestion
Sometimes the commit itself isn't the problem, but rather that it
```

**Review by nwf — COMMENTED** [review]
<3  Thank you for taking the time to write this down.  Some minor nits 's all.

File: CHERI-UPDATING.md (hunk 2 above)
Comment (bsdjhb) [reply CHERI-UPDATING.md:107]: My guess is `git config merge.renames false`

FWIW, I did this as a local config in the checkout vs doing it globally.

**Review by nwf — APPROVED** [review]

File: CHERI-UPDATING.md [hunk 5]
```diff
@@ -79,6 +79,155 @@ Each merge from upstream FreeBSD to [dev] is accompanied by a tag of the form
...
+sure to indicate that the fixup commit fixes an issue introduced in a
+particular commit to aid future users of bisect.
+
+#### Finishing up
+Once the merge is complete and basic local testing is done, it's time to
+update the `freebsd-main` branch and push the tag. (It's good to delay
+to this point in hopes of keeping `freebsd-main` relatively in sync with
+main.)
+```
+$ git push origin freebsd-main-<YYYYMMDD>:freebsd-main
```
Comment (bsdjhb) [inline CHERI-UPDATING.md:198]: Sadly, this command did not work and I had to resort to checking out `freebsd-main`, merging the change, and then pushing `freebsd-main`.  This seems like maybe a bug in GH?
```suggestion
$ git checkout freebsd-main
$ git merge freebsd-main-<YYYYMMDD>
$ git push
```

Comment (nwf) [reply CHERI-UPDATING.md:198]: Was it complaining that `freebsd-main-YYYYMMDD` wasn't a fast-forward from `freebsd-main`?

Comment (bsdjhb) [reply CHERI-UPDATING.md:198]: No, it's some other error about "bad ref" or the like.  I've lost it in my scroll back.  I'll post it when I do the next merge.

Comment (bsdjhb) [reply CHERI-UPDATING.md:198]: Found the error in Slack scroll back:

```
> git push origin freebsd-main-20210305:freebsd-main
Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
To github.com:CTSRD-CHERI/cheribsd.git
 ! [remote rejected]           freebsd-main-20210305 -> freebsd-main (bad ref)
error: failed to push some refs to 'github.com:CTSRD-CHERI/cheribsd.git'
```

Comment (bsdjhb) [reply CHERI-UPDATING.md:198]: I think this is the only thing that might need fixing before merging this in?

Comment (brooksdavis) [reply CHERI-UPDATING.md:198]: Yeah, I'm still confused as to why this doesn't work. We might just want to suggest a worktree.

Comment (bsdjhb) [reply CHERI-UPDATING.md:198]: I'm effectively using a worktree currently.

Comment (brooksdavis) [reply CHERI-UPDATING.md]: I've restored the worktree based text from the original commit.

File: CHERI-UPDATING.md [hunk 6]
```diff
@@ -79,6 +79,156 @@ Each merge from upstream FreeBSD to [dev] is accompanied by a tag of the form
...
+and restarting can make sense.
+
+Sometimes there are too many awkward changes (e.g. one that required
+fixup in files outside the original change) and it's not worth
+backtracking to fix things in the ideal location. In that case, make
+sure to indicate that the fixup commit fixes an issue introduced in a
+particular commit to aid future users of bisect.
+
+#### Finishing up
+Once the merge is complete and basic local testing is done, it's time to
```
Comment (bsdjhb) [inline CHERI-UPDATING.md:193]: One note I guess is that I've been deferring doing these steps until the merge PR completes successfully and push all three things (new tag, freebsd-main, dev) at once.

**Pull request merged**